components/ShadowLayer/src/MDCShadowLayer.m

```diff
@@ -360,6 +367,78 @@ - (void)commonLayoutSublayers {
+
+  UIBezierPath *currentMaskPath = [self outerMaskPath];
+  [currentMaskPath appendPath:currentLayerPath];
+  [currentMaskPath setUsesEvenOddFillRule:YES];
+
+  UIBezierPath *newMaskPath = [self outerMaskPath];
+  [newMaskPath appendPath:newLayerPath];
+  [newMaskPath setUsesEvenOddFillRule:YES];
+
+  // Animate the top layers
+  NSString *shadowPathKey = @"shadowPath";
+  CABasicAnimation *topLayerAnimation = [CABasicAnimation animationWithKeyPath:shadowPathKey];
+  topLayerAnimation.fromValue = (__bridge id)currentLayerPath.CGPath;
+  topLayerAnimation.toValue = (__bridge id)newLayerPath.CGPath;
+  topLayerAnimation.duration = duration;
+  topLayerAnimation.timingFunction = timingFunction;
+  [self.topShadow addAnimation:topLayerAnimation forKey:shadowPathKey];
+  CABasicAnimation *bottomLayerAnimation = [CABasicAnimation animationWithKeyPath:shadowPathKey];
+  bottomLayerAnimation.fromValue = (__bridge id)currentLayerPath.CGPath;
+  bottomLayerAnimation.toValue = (__bridge id)newLayerPath.CGPath;
+  bottomLayerAnimation.duration = duration;
+  bottomLayerAnimation.timingFunction = timingFunction;
+  [self.bottomShadow addAnimation:bottomLayerAnimation forKey:shadowPathKey];
+
+  // Animate the masks
+  if (self.shadowMaskEnabled) {
+    NSString *pathKey = @"path";
+    CABasicAnimation *topMaskLayerAnimation = [CABasicAnimation animationWithKeyPath:pathKey];
+    topMaskLayerAnimation.fromValue = (__bridge id)currentMaskPath.CGPath;
+    topMaskLayerAnimation.toValue = (__bridge id)newMaskPath.CGPath;
+    topMaskLayerAnimation.duration = duration;
+    topMaskLayerAnimation.timingFunction = timingFunction;
+    [self.topShadowMask addAnimation:topMaskLayerAnimation forKey:pathKey];
+    CABasicAnimation *bottomMaskLayerAnimation = [CABasicAnimation animationWithKeyPath:pathKey];
+    bottomMaskLayerAnimation.fromValue = (__bridge id)currentMaskPath.CGPath;
+    bottomMaskLayerAnimation.toValue = (__bridge id)newMaskPath.CGPath;
+    bottomMaskLayerAnimation.duration = duration;
+    bottomMaskLayerAnimation.timingFunction = timingFunction;
+    [self.bottomShadowMask addAnimation:bottomMaskLayerAnimation forKey:pathKey];
+  }
+
+  // Animate the corner radius
+  CABasicAnimation *cornerRadiusAnimation = [CABasicAnimation animationWithKeyPath:@"cornerRadius"];
+  cornerRadiusAnimation.fromValue = @((CGFloat)currentCornerRadius);
+  cornerRadiusAnimation.toValue = @((CGFloat)newCornerRadius);
+  cornerRadiusAnimation.duration = duration;
+  cornerRadiusAnimation.timingFunction = timingFunction;
+  [self addAnimation:cornerRadiusAnimation forKey:@"cornerRadius"];
+
+  // Set completion block
+  [CATransaction setAnimationDuration:duration];
+  [CATransaction setCompletionBlock:^(void) {
+    self.topShadow.shadowPath = newLayerPath.CGPath;
+    self.bottomShadow.shadowPath = newLayerPath.CGPath;
+    self.topShadowMask.path = newMaskPath.CGPath;
+    self.bottomShadowMask.path = newMaskPath.CGPath;
+    self.cornerRadius = newCornerRadius;
```

Comment:
When using CAAnimation objects it's good practice to, in nearly all cases, **set the model value to the animation's destination**. This removes the need for completion blocks like this one that fix the "snapping" behavior because our layer will already reflect the desired destination value.

The reason this works is because so long as an animation exists on a layer for a given key path, iOS will prefer showing the **computed** value of the animation. The inverse applies as well: once there are no animations for a given key path, iOS will start rendering the **model** value of the layer again. If the model value wasn't updated to the destination, then we see the classic "snap" effect that plagues so much explicit Core Animation code.

Additional fun fact: this is why the UIView implicit animation APIs work the way they do! When you set a view's property in an animation block, Core Animation will generate a CAAnimation from the property's original value to the new one and then commit the new value to model layer.

Reply:
With that in mind, do you think alternative 2 is a better API?